offer-manager-domain/src/main/groovy/life/qbic/business/products/copy/CopyProduct.groovy

```diff
@@ -0,0 +1,134 @@
+import life.qbic.business.products.create.CreateProductInput
+import life.qbic.business.products.create.CreateProductOutput
+import life.qbic.datamodel.dtos.business.services.Product
+import org.aspectj.bridge.IMessage
+
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+
+/**
+ * <h1>4.3.2 Copy Service Product</h1>
+ * <br>
+ * <p> Offer Administrators are allowed to create a new permutation of an existing product.
+ * <br> New permutations can include changes in unit price, sequencing technology and other attributes of service products.
+ * </p>
+ *
+ * @since: 1.0.0
+ *
+ */
+class CopyProduct implements CopyProductInput, CreateProductOutput {
+
+    private static final Logging log = Logger.getLogger(this.class)
+
+    private final CopyProductDataSource dataSource
+    private final CopyProductOutput output
+    private final CreateProductInput createProduct
+
+    /**
+     * The only constructor for this use case
+     * @param dataSource - a data source that provides mandatory functionality
+     * @param output - an output that provides mandatory functionality
+     * @param createProductInput - a CreateProduct use case that is used to create the product
+     */
+    CopyProduct(CopyProductDataSource dataSource, CopyProductOutput output, CreateProductDataSource createProductDataSource) {
+        this.dataSource = dataSource
+        this.output = output
+        this.createProduct = new CreateProduct(createProductDataSource, this)
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    void copyModified(Product product) {
+        Optional<Product> searchResult
+        try{
+            //1. retrieve product from db
+            searchResult = dataSource.fetch(product.getProductId())
+            if (searchResult.isPresent()) {
+                Product existingProduct = searchResult.get()
+                //2. compare if there is a difference between the products in order
+                if(getProductChecksum(product) != getProductChecksum(existingProduct)){
+                    //3. call the CreateProduct use case (new id is created here)
+                    createProduct.create(product)
+                }else{
+                    foundDuplicate(product)
+                }
+            }else{
+                //there is no product present, this should not happen
+                log.error("An unexpected during the project creation occurred.")
+                output.failNotification("An unexpected during the project creation occurred. " +
+                        "Please contact ${Constants.QBIC_HELPDESK_EMAIL}.")
+            }
+        }catch(DatabaseQueryException databaseQueryException){
+            log.error("The copied product ${product.productId.toString()} cannot be found in the database", databaseQueryException)
+            output.failNotification("The copied product ${product.productId.toString()} cannot be found in the database")
```

Comment:
I think the readability can be improved here. Clean Code suggests to use a private method, where you then export the code block that will through an exception, and call this in a try/catch from the top method level.